### PR TITLE
[SMTChecker] Add mod operator

### PR DESCRIPTION
**Pull request by leonardoalt**

### CODE REVIEW

**Comment by axic** [comment]
Are we tracking smt features in the changelog?

**Comment by axic** [comment]
Well this one doesn't even seem to be exposed to language features,

**Comment by leonardoalt** [comment]
We are, but this is just an internal change to enable the usage in the SMTChecker which will be in the Changelog.

**Comment by leonardoalt** [comment]
It looks like the `z3` version used in the tests doesn't have `mod`. Trying to figure what version introduced it. My local 4.8.0 has it.

File: libsolidity/formal/SolverInterface.h [hunk 1]
```diff
@@ -141,6 +141,7 @@ class Expression
 			{"-", 2},
 			{"*", 2},
 			{"/", 2},
+			{"mod", 2},
```
Comment (erak) [inline libsolidity/formal/SolverInterface.h:144]: Can these be sorted by arity? Or alphabetically? Not sure which I'd prefer ;) But maybe by arity makes more sense. This sorting should then also be applied for the operator overloadings.

Comment (leonardoalt) [reply libsolidity/formal/SolverInterface.h:144]: For now it's sort of sorted by theories: first Boolean operators, then arithmetic operators, then array operators.

Comment (erak) [reply libsolidity/formal/SolverInterface.h:144]: Ah, I see. Thanks!

**Comment by leonardoalt** [comment]
Z3 added `mod` in version 4.6.0.
Ubuntu Cosmic uses Z3 4.4, therefore this won't work in the tests.

I asked Z3 directly and they are not maintaining it nor will.
I asked directly in the ubuntu package page if there are update plans (https://answers.launchpad.net/ubuntu/+source/z3/+question/677999).

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
What happens if I try to compile this with an older version of z3?

**Comment by leonardoalt** [comment]
It will fail to compile saying `z3::mod` doesn't exist.

**Comment by chriseth** [comment]
Is it possible to add a CMAKE check to force a certain version?
My aim is to have a message that tells users which version to install.

**Comment by leonardoalt** [comment]
Ok will try that.

**Comment by axic** [comment]
@leonardoalt didn't you say no Ubuntu version has an appropriate z3? Hence releases cannot be made with z3 anymore? (Or they aren't?)

If we do make releases with z3 I'd opt to have an ifdef and warning at runtime if using that feature. If we do not care about z3 support in releases, I'm more than happy to merge as is, given the cmake-time version check.

**Comment by leonardoalt** [comment]
@axic After https://github.com/ethereum/solidity/pull/5897 was merged I take it that releases will not have z3 anymore. I agree with your second option.

**Comment by chriseth** [comment]
Hm - in that case, is it really not possible to have a workaround, i.e. to define mod manually?

**Comment by leonardoalt** [comment]
I'm not sure that'd be a good workaround. One workaround would be to not use `mod` at all, but if we do I'd stick to the correct operator.

**Comment by leonardoalt** [comment]
Since I removed the `QUIET` option, CMake will inform whatever happens to finding Z3. I can also keep `QUIET` and use an if instead of `VERSION_VAR` to report that the version is wrong.

**Review by chriseth — APPROVED** [review]

**Review by chriseth — CHANGES_REQUESTED** [review]
Postponed merging until the PPA sitution is clarified.

**Comment by chriseth** [comment]
PPA change: https://github.com/ethereum/solidity/pull/5945

**Review by chriseth — APPROVED** [review]

**Pull request merged**